src/pages/VehicleRegister/VehicleRegister.jsx

```diff
@@ -0,0 +1,192 @@
+import React, { useEffect, useState, useCallback } from "react";
+import { useHistory, useParams } from "react-router";
+import { Button, TextField, Box, FormHelperText } from "@material-ui/core";
+import { Select } from "../../components";
+import useFormValidations from "../../hooks/useFormValidations";
+import VehicleService from "../../services/VehicleService";
+import BrandService from "../../services/BrandService";
+import {
+  composeValidators,
+  validYear,
+  required,
+  onlyNumbers,
+  minValue,
+  minLength,
+} from "../../utils/validations/validations";
+
+const validations = {
+  model: (value) => {
+    return minLength(2, "Modelo deve ter ao menos 2 letras.")(value);
+  },
+  year: (value) => {
+    return composeValidators(required(), onlyNumbers(), validYear())(value);
+  },
+  price: (value) => {
+    return composeValidators(required(), onlyNumbers(), minValue(1))(value);
+  },
+  brand: (value) => {
+    return required("Selecione uma marca")(value);
+  },
+};
+
+function VehicleRegister() {
+  const { id } = useParams();
+  const history = useHistory();
+
+  const [brandId, setBrandId] = useState("");
+  const [model, setModel] = useState("");
+  const [year, setYear] = useState("");
+  const [price, setPrice] = useState("");
+  const [brands, setBrands] = useState([]);
+  const [error, setError] = useState("");
+  const [errors, validateFields, shouldSubmit] =
+    useFormValidations(validations);
+
+  function cancel() {
+    history.goBack();
+  }
+
+  function handleSubmit(event) {
+    event.preventDefault();
+    if (shouldSubmit()) {
+      submitBrand();
+    }
+  }
+
+  async function submitBrand() {
+    try {
+      setError("");
+      if (id) {
+        await VehicleService.update({
+          id,
+          model,
+          year,
+          price,
+          brandId,
+        });
+        return history.push("/");
+      }
+      await VehicleService.register({
+        model,
+        year,
+        price,
+        brandId,
+      });
+      return history.push("/");
+    } catch (e) {
+      setError(e.data);
+    }
+  }
+
+  const loadVehicleFromId = useCallback(async () => {
+    if (id) {
+      try {
+        const {
+          data: { model, year, price, brandId },
+        } = await VehicleService.get(id);
+        setModel(model);
+        setYear(year);
+        setPrice(price);
+        setBrandId(brandId);
+      } catch (e) {
+        setError(e.data);
+      }
+    }
+  }, [id]);
+
+  useEffect(() => {
+    loadVehicleFromId();
+  }, [loadVehicleFromId]);
+
+  const loadBrands = useCallback(async () => {
+    try {
+      const brands = await BrandService.getAll();
+      return setBrands(brands);
+    } catch (e) {
+      setError(e.data);
+    }
+  }, [setBrands]);
```

Comment:
Acho que esse `brands` ser declarado com `useMemo`, porque ele só é setado dentro do contexto do `useCallback`.

```suggestion
  const brands = useMemo(async () => {
    try {
      return await BrandService.getAll();
    } catch (e) {
      setError(e.data);
      return [];
    }
  }, [setBrands]);
```

Reply:
eu nao entendi pq usar useMemo aqui.. Pode esclarecer? @WaaLopes

Reply:
É que você criou o `const [brands, setBrand] = useState`, e ainda precisou de um `useEffect` com o `useCallback`, só pra definir o valor do `brands`. Acho que a implementação com o `useMemo` faz a mesma coisa, só que de maneira mais direta.

Reply:
vc ja usou essa estrategia, @WaaLopes? pq fui tentar e o useMemo retorna uma Promise, não tem como guardar o retorno dele dentro de uma variável

alem de que ele geralmente é usado pra lidar com processamento custosos... o que eu pensei é setar o brands dentro da variavel usando o useMemo, no entanto, eu estaria usando um anti pattern (eu acho)

Reply:
tentei isso -> 

```js
```

Reply:
> vc ja usou essa estrategia, @WaaLopes? pq fui tentar e o useMemo retorna uma Promise, não tem como guardar o retorno dele dentro de uma variável

Eu usava algo parecido com o Apollo Client, e pensei que o `useMemo` teria uma maneira de lidar com esse resultado async e atualizar ao resolver a promise, mas parece que não rola mesmo. My bad =/

> tentei isso ->
> 
> ```js
>   useEffect(() => {
>     const loadBrands = async () => {
>       try {
>         const brands = await BrandService.getAll();
>         return setBrands(brands);
>       } catch (e) {
>         setError(e.data);
>       }
>     };
> 
>     loadBrands();
>   });
> ```
> 
> Mas deu um problema de leak do useEffect(). Acho que precisaria dar um return para limpar essa operação, algo assim... Mas não sei se vale a pena :x

Eu gostei dessa forma sem o `useCallback`, já que o `loadBrands` só é necessário nesse contexto mesmo.
O leak do `useEffect` ocorreu porque tu não passou um array vazio como segundo parâmetro.